src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -275,45 +310,45 @@ internal static unsafe long ParseNonCanonical(char* name, int start, ref int end
             {
                 // end of string, allowed
             }
-            else if ((ch = name[current]) == '/' || ch == '\\' || (notImplicitFile && (ch == ':' || ch == '?' || ch == '#')))
+            // For a normal IPv4 address, the terminator is the prefix ('/' or its counterpart, '\'). If notImplicitFile is set, the terminator
+            // is one of the characters which signify the start of the rest of the URI - the port number (':'), query string ('?') or fragment ('#')
+            else if (name[current] == TChar.CreateTruncating('/') || name[current] == TChar.CreateTruncating('\\')
+                    || (notImplicitFile && (name[current] == TChar.CreateTruncating(':') || name[current] == TChar.CreateTruncating('?') || name[current] == TChar.CreateTruncating('#'))))
             {
-                end = current;
+                // end of string, (as terminated) allowed
             }
             else
             {
                 // not a valid terminating character
                 return Invalid;
             }
 
-            parts[dotCount] = currentValue;
+            parts[dotCount] = unchecked((uint)currentValue);
+            end = current;
 
-            // Parsed, reassemble and check for overflows
+            // Parsed, reassemble and check for overflows in the last part. Previous parts have already been checked in the loop
             switch (dotCount)
             {
                 case 0: // 0xFFFFFFFF
-                    if (parts[0] > MaxIPv4Value)
-                    {
-                        return Invalid;
-                    }
                     return parts[0];
                 case 1: // 0xFF.0xFFFFFF
                     if (parts[1] > 0xffffff)
                     {
                         return Invalid;
                     }
-                    return (parts[0] << 24) | (parts[1] & 0xffffff);
+                    return (parts[0] << 24) | parts[1];
```

Comment:
The masks are no longer necessary?

Reply:
At the point of processing each `.` in an IPv4 address, the parsing logic exits if the component's value is greater than 0xFF. The final IPv4 component's range is checked in each case, so the masks will always effectively be a NOP.